### PR TITLE
feat(deploy): UI based deploy flow

### LINKED ISSUE #5216: Push credentials from local to cloud as part of deploy flow
For connectors whose credentials are explicitly specified as part of the connector configuration in the .env file. We would like to push those credentials as part of the project deploy, In order to avoid any errors after the deployment. 

The admin would have the option to change the credentials via `rill env configure` and via admin UI in the future. 

This is a subtask of the bigger credential management.

### PR DESCRIPTION
**Pull request by AdityaHegde**
closes [#4969](https://github.com/rilldata/rill/issues/4969) closes #5216

Testing steps,
1. Create cli with this: `make cli`
2. Start cloud (needs docker running): `./rill devtool start cloud`
3. (Optional) Copy cli to home `cp ./rill ~/rill` and remove quarantine: `sudo xattr -d com.apple.quarantine ~/rill`
4. Run this within a new folder or existing folder: `~/rill start`

This will work for non-github and github based projects. Redeploy wont show for projects deployed before this.

### CODE REVIEW

**Comment by ericokuma** [comment]
Hey Aditya! When I click on Deploy, I get a 500 error
Screenshot 2024-07-02 at 9 23 28 AM

File: cli/pkg/local/server.go
Comment (k-anshul) [inline cli/pkg/local/server.go:196]: We can end up in a state where `isGithubRepo` is false but `githubRemoteFound` is true which doesn't seem correct.

File: cli/pkg/local/server.go
Comment (k-anshul) [inline cli/pkg/local/server.go:148]: I think these two flags can be avoided by keeping a `var p *Project` directly ?

File: proto/rill/local/v1/api.proto
Comment (k-anshul) [inline proto/rill/local/v1/api.proto:32]: I don't know the full e2e flow from UI but for CLI we are mostly creating first and if conflict exist then prompting for project/org names. The org name can still be taken if this check passes plus it avoids extra network call.

**Review by k-anshul — CHANGES_REQUESTED** [review]

File: cli/pkg/local/server.go
Comment (AdityaHegde) [reply cli/pkg/local/server.go:196]: Hmm this code was moved around. But I agree it doesnt make sense.

File: proto/rill/local/v1/api.proto
Comment (AdityaHegde) [reply proto/rill/local/v1/api.proto:32]: What could be reason for org name to exist even if this passes? Race condition?

We do not have a flow to prompt user in UI. It has to automatically select the next available by adding `-1`, `-2` suffixes.

Comment (k-anshul) [reply proto/rill/local/v1/api.proto:32]: > What could be reason for org name to exist even if this passes? Race condition?

Yes

> It has to automatically select the next available by adding -1, -2 suffixes.

Then I would suggest lets create first and if it fails due to name mismatches then select the next available ?

File: web-local/src/routes/(misc)/deploy/+page.svelte
Comment (ericpgreen2) [inline web-local/src/routes/(misc)/deploy/+page.svelte:54]: Here too we should use `href` not `on:click` for links

File: web-common/src/features/project/ProjectDeployer.ts
Comment (ericpgreen2) [inline web-common/src/features/project/ProjectDeployer.ts:83]: It initially surprised me to see the `validate()` function opens up a log-in page as a side effect. Might be more clear if you moved this to the `deploy()` function.

File: web-admin/vite.config.ts
Comment (ericpgreen2) [inline web-admin/vite.config.ts:37]: Can we move this function elsewhere?

File: web-admin/src/routes/[organization]/[project]/-/invite/+page.svelte
Comment (ericpgreen2) [inline web-admin/src/routes/[organization]/[project]/-/invite/+page.svelte:21]: Missing error handling for these queries & mutation

File: web-common/src/features/project/ProjectDeployer.ts
Comment (ericpgreen2) [inline web-common/src/features/project/ProjectDeployer.ts:129]: With the `try/catch` and 3 nested `if` statements, this is pretty deep indentation. Consider if there's a way to avoid such nesting, e.g. with guard clauses or factoring out a `getOrg()` function.

**Review by ericpgreen2 — CHANGES_REQUESTED** [review]
UXQA: the `/deploy` page's loading spinner flashes with a "Need help" message. See [Jam video](https://jam.dev/c/c01b71c3-74ba-4988-9c99-4487951ad0b8).

**Review by k-anshul — APPROVED** [review]

File: web-common/src/features/project/deploy-errors.ts
Comment (AdityaHegde) [reply web-common/src/features/project/deploy-errors.ts:19]: Not right now. We can do a follow up to handle this in backend and send a consistent error upfront. I believe there is a way to pre calculate the size.

File: web-local/src/routes/(misc)/deploy/+page.svelte
Comment (AdityaHegde) [reply web-local/src/routes/(misc)/deploy/+page.svelte:54]: This will open pylon similar to AvatarButton. Need pylon integration to local before this is merged.

**Comment by ericpgreen2** [comment]
We should avoid this parse error, and anyways we should probably not show the `.rillcloud` directory in the file explorer. Can we add `.rillcloud` to the default `ignore_paths`? Or will that prevent the runtime from writing to the directory?

image

image

File: web-common/src/features/dashboards/workspace/DeployDashboardCTA.svelte
Comment (ericpgreen2) [inline web-common/src/features/dashboards/workspace/DeployDashboardCTA.svelte:58]: These are reversed

File: web-admin/src/lib/vite-utils.ts
Comment (ericpgreen2) [inline web-admin/src/lib/vite-utils.ts:1]: Sorry, I could have been more clear – this function should live with the "feature" that it relates to. Looks like it relates to the code in `features/projects/user-invite`

File: web-local/src/routes/(misc)/deploy/+page.svelte
Comment (ericpgreen2) [reply web-local/src/routes/(misc)/deploy/+page.svelte:55]: Oh now I see. Given the event handler isn't yet implemented, we should hide the button.

File: web-common/src/features/project/OrgSelector.svelte
Comment (ericpgreen2) [inline web-common/src/features/project/OrgSelector.svelte:20]: Nit: trailing slash

File: web-admin/src/routes/[organization]/[project]/-/invite/+page.svelte
Comment (ericpgreen2) [inline web-admin/src/routes/[organization]/[project]/-/invite/+page.svelte:49]: We should only continue to the `/status` page if there's no error. Rather than show an ephemeral notification, we should show the error in red text below the "Continue" button.

File: web-local/src/routes/(misc)/deploy/+page.svelte
Comment (ericpgreen2) [inline web-local/src/routes/(misc)/deploy/+page.svelte:50]: No need for this `h-36` – it adds funky extra space in this case

File: web-admin/src/routes/[organization]/[project]/-/invite/+page.svelte
Comment (AdityaHegde) [reply web-admin/src/routes/[organization]/[project]/-/invite/+page.svelte:49]: I disagree, error while adding a domain whitelist should not block the user from going forward. This can barely happen and there are no mocks for this. Lets do a follow up for this.

Comment (ericpgreen2) [reply web-admin/src/routes/[organization]/[project]/-/invite/+page.svelte:49]: Sounds good, I won't block on this. @ericokuma for resolution.

File: web-common/src/features/project/ProjectDeployer.ts
Comment (ericpgreen2) [inline web-common/src/features/project/ProjectDeployer.ts:64]: I think this is a private function, not called externally.

**Review by ericpgreen2 — APPROVED** [review]
Approving from the frontend perspective. Great feature!

File: web-common/src/features/project/ProjectDeployer.ts
Comment (AdityaHegde) [reply web-common/src/features/project/ProjectDeployer.ts:79]: Ah the validate method is meant to read through DeployValidation and make sure it passes.

**Review by ericokuma — APPROVED** [review]
looks good

**Pull request merged**
